### DISCUSSION TITLE
Any way to enable lsp for only the selectedTextEditor?

### DISCUSSION THREAD

**Discussion by onriv** [redhat-developer/lsp4ij#582] [General]
Hi in our plugin the lsp server  is some kind performance intensive. We want to see if it's possible limit the lsp server request only for the current focus file, like https://github.com/onriv/lean4ij/issues/35. Checking the requests our lsp server start some heavy operations once the `org.eclipse.lsp4j.services.TextDocumentService#didOpen` requests is sent.

Is there any way for configuring this? The `com.redhat.devtools.lsp4ij.client.features.LSPClientFeatures` seems not suitable here. And checking the code [LanguageServerWrapper#connect](https://github.com/redhat-developer/lsp4ij/blob/main/src/main/java/com/redhat/devtools/lsp4ij/LanguageServerWrapper.java#L612) seems not help either.

**Comment by angelozerr** [comment]
Perhaps by overriding all LSP*feature#isEnabled and get the editor from PsiFile and check if the editor is the current editor?

**Reply by onriv** [reply]
Yeah that was what I planned to do but I found that there is no LSP*feature that for sending didOpen request. The server end seems do the heavy job while handling this didOpen. Let me check if the server end can do anything

**Reply by angelozerr** [reply]
If I understand your need, you would like forbid some didOpen?

If it that you would like to have a method LSPClientFeatures#didOpen(PsiFile file) that you would like to override to avoid sending the real LSP didOpen request, is that?

**Reply by onriv** [reply]
Yeah indeed. Debugging I found that for the non-slectedfile:
![image](https://github.com/user-attachments/assets/621469b5-97ab-4c29-a137-9b464c616770)
There is still some requests sending to the server like:
![image](https://github.com/user-attachments/assets/aef926e4-7b30-48d4-98b6-ecb00aa83489)

**Reply by onriv** [reply]
Maybe in the case for lean it's kind of prefering to have an option that only after taking the file to front, it's treated as open and sends the `didOpen` request

**Reply by angelozerr** [reply]
LSP4IJ sends a didOpen for all opened editor. We could move the logic of didOpen in LSPClientFeatures and even add a method like shouldConnect(PsiFile file) that you could override it to check if didOpen must be sent.

The main problem is that didOpen is mapped with fileOpened event listener which is called for all opened editor.

We should track for your case when editor is selected.

**Reply by onriv** [reply]
Yeah checking the code. Is the suitable extension point [here](https://github.com/redhat-developer/lsp4ij/blob/main/src/main/java/com/redhat/devtools/lsp4ij/LanguageServiceAccessor.java#L248) here? 

I am not sure if it's possible turning into things like
```
if (!wrapper.isEnabled(file)) {
    return CompletableFuture.completedFuture(Collections.emptyList());
}
if (server != null &&
        (afterStartingServerFilter == null || afterStartingServerFilter.test(wrapper.getClientFeatures()))) {
    return wrapper.connect(file, document);
}
return CompletableFuture.completedFuture(null);
```

and make `wrapper.isEnable(file)` an extension point.

**Reply by onriv** [reply]
BTW there is a similar case in our situation: when poping up searching dialog, the file is also treated as opened and send the `didOpen` request. For avoiding this while implementing the factory `LanguageServerFactory`, the `LanguageServerFactory#isEnabled(project)` method turns true if only the real editor is focused. But it has no file argument now.

**Comment by angelozerr** [comment]
@onriv I have a question. Your performance problem occurs when IJ is starting and opens previously opened editors which starts your ls  ? 

If it that see https://github.com/redhat-developer/lsp4ij/discussions/571

Or have you this problem again when IJ is started and you open qeveral editors?

**Reply by onriv** [reply]
> @onriv I have a question. Your performance problem occurs when IJ is starting and opens previously opened editors which starts your ls ?
> 
> If it that see #571
> 
> Or have you this problem again when IJ is started and you open qeveral editors?

It seems neither. It seems that our language server is indeed cpu intensive for first opening the file:
<img width="640" alt="image" src="https://github.com/user-attachments/assets/400f4d62-9da7-4af6-8923-9b4d4c11c430">
and hence it may help for only handling current focused editor.

I made a PR minutes ago, please check if it's suitable?
https://github.com/redhat-developer/lsp4ij/pull/586

**Comment by onriv** [comment]
closed with https://github.com/redhat-developer/lsp4ij/pull/586
